presto-main/src/main/java/io/prestosql/type/UuidOperators.java

```diff
@@ -60,6 +67,123 @@ public static Slice uuid()
         return wrappedLongArray(uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
     }
 
+    @Description("Returns the nil UUID constant, which does not occur as a real UUID")
+    @ScalarFunction
+    @SqlType(StandardTypes.UUID)
+    public static Slice uuid_nil()
+    {
+        java.util.UUID nilUuid = new UUID(0L, 0L);
+        return wrappedLongArray(nilUuid.getMostSignificantBits(), nilUuid.getLeastSignificantBits());
+    }
+
+    @Description("Returns the DNS namespace constant for UUIDs")
+    @ScalarFunction
+    @SqlType(StandardTypes.UUID)
+    public static Slice uuid_ns_dns()
+    {
+        String nameSpaceDns = "6ba7b810-9dad-11d1-80b4-00c04fd430c8";
+        java.util.UUID nsDnsUuid = java.util.UUID.fromString(nameSpaceDns);
+        return wrappedLongArray(nsDnsUuid.getMostSignificantBits(), nsDnsUuid.getLeastSignificantBits());
+    }
+
+    @Description("Returns the URL namespace constant for UUIDs")
+    @ScalarFunction
+    @SqlType(StandardTypes.UUID)
+    public static Slice uuid_ns_url()
+    {
+        String nameSpaceUrl = "6ba7b811-9dad-11d1-80b4-00c04fd430c8";
+        java.util.UUID nsUrlUuid = java.util.UUID.fromString(nameSpaceUrl);
+        return wrappedLongArray(nsUrlUuid.getMostSignificantBits(), nsUrlUuid.getLeastSignificantBits());
+    }
+
+    @Description("Returns the ISO object identifier (OID) namespace constant for UUIDs")
+    @ScalarFunction
+    @SqlType(StandardTypes.UUID)
+    public static Slice uuid_ns_oid()
+    {
+        String nameSpaceOid = "6ba7b812-9dad-11d1-80b4-00c04fd430c8";
+        java.util.UUID nsOidUuid = java.util.UUID.fromString(nameSpaceOid);
+        return wrappedLongArray(nsOidUuid.getMostSignificantBits(), nsOidUuid.getLeastSignificantBits());
+    }
+
+    @Description("Returns the X.500 distinguished name (DN) namespace constant for UUIDs")
+    @ScalarFunction
+    @SqlType(StandardTypes.UUID)
+    public static Slice uuid_ns_x500()
+    {
+        String nameSpaceX500 = "6ba7b814-9dad-11d1-80b4-00c04fd430c8";
+        java.util.UUID nsX500Uuid = java.util.UUID.fromString(nameSpaceX500);
+        return wrappedLongArray(nsX500Uuid.getMostSignificantBits(), nsX500Uuid.getLeastSignificantBits());
+    }
+
+    @Description("Generates a deterministic (type 3) UUID using the specified namespace UUID and input name (using MD5 hash)")
+    @ScalarFunction
+    @SqlType(StandardTypes.UUID)
+    public static Slice uuid_v3(@SqlType(StandardTypes.UUID) Slice nameSpaceUuid, @SqlType(StandardTypes.VARCHAR) Slice name)
+    {
+        byte[] value = name.getBytes();
+
+        java.util.UUID nameUuid = new java.util.UUID(nameSpaceUuid.getLong(0), nameSpaceUuid.getLong(SIZE_OF_LONG));
+        String uuidHexString = nameUuid.toString().replace("-", "");
+
+        byte[] nameSpaceBytes = new byte[16];
+        for (int i = 0; i < 16; i++) {
+            int firstDigit = Character.digit(uuidHexString.substring(i * 2, i * 2 + 2).charAt(0), 16);
```

Comment:
If you go with parsing using `parseInt` is simpler:
```java
byte b = (byte) parseInt(uuidHexString.substring(i * 2, i * 2 + 2), 16);
```

But maybe directly operating on:
```java
        long mostSigBits = uuid.getLong(0);
        long leastSigBits = uuid.getLong(SIZE_OF_LONG);
```
would be more readable/more performant.

@martint is it ok to exploit how we structure the UUID in `Slice` and treat it as an external contract? Or should it be treated as an internal (opaque) representation, where we only know that we can feed it to `UUID(long, long)` constructor?

Reply:
Yes, functions and operators have full visibility into how the uuid is represented in the Slice.

Reply:
> Yes, functions and operators have full visibility into how the uuid is represented in the Slice.

@gamebusterz then I suggest that we avoid conversion to string and directly operate on binary representation of namespace UUID.

Reply:
@losipiuk I'm not sure, but I think this is what electrum suggested https://github.com/prestosql/presto/pull/4374#issuecomment-655738847 

This approach doesn't give the same UUIDs (for `uuid_v3` and `uuid_v5` only, the constant UUID functions `ns_dns`,`ns_url` etc. work as expected) as other implementations. The resultant UUID will be deterministic, but it'll be different from what `uuid-ossp` (Postgres) or Python implementation of `uuid` would return. 
```
select uuid_v3(uuid_ns_dns(),'presto')
python '8efeec8c-1d77-3a5a-b1bb-22e702970750' 
presto 'b86f9a30-3fb8-9a50-a4ab-854fc0f9d68f' <----different
psql   '8efeec8c-1d77-3a5a-b1bb-22e702970750'
```
The converting to string approach gave the expected results. 

Let me know if my understanding is wrong, I'm willing to try out any suggestions

Reply:
The layout of the uuid inside the Slice is in little-endian form, so you need to account for that. The function should produce the same result as python's or postgres' implementation.